### PR TITLE
feat(resource-claims): Optimistic locking and capacity aggregation [patch-3]

### PR DESCRIPTION
**Pull request by joeysabs**
## What Does this PR Do?

Refactors resource claim system to use optimistic locking instead of row-level locking. Introduces capacity aggregation across all active subscription item features and changes claims querying from subscriptionItemFeatureId to (subscriptionId, resourceId) tuple. This patch stacks on top of PR #1520 which removed subscriptionItemFeatureId from the schema.

---
## Summary by cubic
Switch resource claims to optimistic locking with atomic batch inserts and aggregate capacity across all active features. Claims and usage are now keyed by (subscriptionId, resourceId), and usage endpoints batch capacity/count queries to avoid N+1 calls.

- **Refactors**
  - Aggregate capacity from active features; stop using subscriptionItemFeatureId in queries.
  - Implement optimistic locking with atomic batched compare-and-swap and retry for claim inserts (all-or-nothing).
  - Add getAggregatedResourceCapacityBatch and update router to batch capacity and claim counts (fix N+1 in list usages).
  - Update claimResourceTransaction and releaseResourceTransaction to use (subscriptionId, resourceId).
  - Simplify getResourceUsage(subscriptionId, resourceId); remove row-level locks.

- **Migration**
  - Replace getResourceUsage(subscriptionId, subscriptionItemFeatureId) with getResourceUsage(subscriptionId, resourceId).
  - Replace countActiveClaimsForSubscriptionItemFeature with countActiveResourceClaims, and countActiveClaimsForSubscriptionItemFeatures with countActiveResourceClaimsBatch.
  - Remove validateResourceCapacityForDowngrade and releaseAllResourceClaimsForSubscriptionItemFeature.
  - Update callers, tests, and seed to drop subscriptionItemFeatureId usage.

<sup>Written for commit 9fcb2978dd5d5dea3b883d0cba5588a742dc2a24. Summary will update on new commits.</sup>

### CODE REVIEW

File: platform/flowglad-next/src/resources/resourceClaimHelpers.ts [hunk 1]
```diff
@@ -318,93 +311,325 @@ const findResourceBySlug = async (
...
+          resourceId: insertedRow.resource_id,
+          subscriptionId: insertedRow.subscription_id,
+          pricingModelId: insertedRow.pricing_model_id,
+          externalId: insertedRow.external_id,
+          metadata: insertedRow.metadata,
+          livemode: insertedRow.livemode,
+          claimedAt: insertedRow.claimed_at,
+          releasedAt: insertedRow.released_at,
+          releaseReason: insertedRow.release_reason,
+        })
```
Comment (agreea) [inline platform/flowglad-next/src/resources/resourceClaimHelpers.ts]: can we zod parse this result to be safe?

**Review by agreea — APPROVED** [review]
good to go pending zod parsing the SQL, and probably removing `console.log` (unless it's dependent on other patterns).
---
## Code Review Summary: PR #1521

### Overview

PR #1521 implements **Patch 3** of the `resource-claim-decoupling` gameplan. It refactors the resource claims system from per-feature locking to optimistic locking with capacity aggregation across all active subscription item features.

### Alignment with Gameplan ✅

The implementation closely follows the gameplan specifications:

| Gameplan Requirement | Implementation Status |
|---------------------|----------------------|
| Remove `subscriptionItemFeatureId` from claim inserts | ✅ Done |
| Add `getAggregatedResourceCapacity` function | ✅ Done |
| Implement optimistic locking with conditional insert | ✅ Done (using UNNEST for atomic batch insert) |
| Update `claimResourceTransaction` to use new approach | ✅ Done |
| Update `releaseResourceTransaction` to query by (subscriptionId, resourceId) | ✅ Done |
| Update `getResourceUsage` signature | ✅ Done |
| Remove `validateResourceCapacityForDowngrade` | ✅ Done |
| Remove `releaseAllResourceClaimsForSubscriptionItemFeature` | ✅ Done |
| Add `getAggregatedResourceCapacityBatch` (N+1 fix) | ✅ Done |

### Issues Identified

**HIGH Severity (1)**
- **Potential Race Condition in Capacity Check** (`resourceClaimHelpers.ts:519`): The optimistic lock guards against concurrent claim count changes but not concurrent capacity changes (e.g., if a subscription item feature expires between capacity read and insert). This is a minor edge case since capacity changes are rare administrative operations.

**MEDIUM Severity (2)**
- **Type Safety Concern** (`resourceClaimHelpers.ts:576`): The result of `transaction.execute()` is cast through `unknown`. Consider using Drizzle's infer types or zod parsing for runtime validation.
- **SQL Template Safety**: Verified - no SQL injection risk. All dynamic values are properly parameterized via Drizzle's `sql` template tag.

**LOW Severity (1)**
- **Console.log in Production** (`resourceClaimHelpers.ts:624`): Consider using a structured logger or removing the retry logging.

### Test Coverage Assessment

**Covered:**
- Atomic batch inserts (all-or-nothing semantics)
- Optimistic locking with no partial inserts
- Full capacity claiming
- Batch counting and capacity aggregation

**Missing (per gameplan Patches 5 & 6):**
- Subscription adjustment survival tests
- Billing run integration tests
- Multi-item capacity aggregation during adjustment tests

These may be intended for subsequent PRs following the gameplan's dependency graph.

### Architecture & Performance

**Positive:**
- Clean removal of feature-scoped methods
- Efficient batch operations to avoid N+1 queries
- Proper use of UNNEST for atomic multi-row inserts
- Consistent (subscriptionId, resourceId) scoping throughout

**Breaking Changes:**
- API: `subscriptionItemFeatureId` removed from ResourceClaim schema
- Behavior: `listResourceUsagesProcedure` now returns one entry per resource with aggregated capacity (instead of one entry per feature)

### Recommendation

**Approve with minor suggestions:**
1. Consider adding zod validation on raw SQL results for runtime safety
2. Replace `console.log` with structured logging
3. Document the capacity race condition edge case in code comments
4. Confirm Patches 5 & 6 tests are planned for follow-up PRs

The implementation is well-aligned with the gameplan and fixes the original problem of claims becoming orphaned after subscription adjustments.
